### PR TITLE
Workflow refactor Phase 2 Part 1

### PR DESCRIPTION
**Pull request by joelridden**
#### Description
New table job_duration_log which stores metadata
Fills in first values of job_id and queued time when a job is queued
Fills in values, start_time, nodes, cores, wct when a task is running
Fills in end time when a task hits failed, killed_WCT or completed
Adjusted templates and generics for Maui/Mahuika and Kisti to add data to the add_to_mgmt_queue.py script as optional parameters

#### Checklist
- Have you updated the README (yes/no)?
- Have you updated the CHANGELOG (yes/no)?

### CODE REVIEW

File: workflow/automation/execution_scripts/add_to_mgmt_queue.py [hunk 1]
```diff
@@ -44,4 +81,24 @@
         const.Status.from_str(args.status).value,
         job_id=args.job_id,
         error=args.error,
+        start_time=int(
+            datetime.strptime(args.start_time, "%Y-%m-%d_%H:%M:%S").timestamp()
+        )
+        if args.start_time is not None
+        else None,
```
Comment (jpa198) [inline workflow/automation/execution_scripts/add_to_mgmt_queue.py:88]: These 3 "a if b else c" statements need temporary variables, this is way too hard to read

File: workflow/automation/execution_scripts/queue_monitor.py [hunk 2]
```diff
@@ -60,6 +60,13 @@ def get_queue_entry(
         status=data_dict[MgmtDB.col_status],
         job_id=data_dict[MgmtDB.col_job_id],
         error=data_dict.get("error"),
+        queued_time=data_dict.get(MgmtDB.col_queued_time),
+        start_time=data_dict.get(MgmtDB.col_start_time),
+        end_time=data_dict.get(MgmtDB.col_end_time),
+        nodes=data_dict.get(MgmtDB.col_nodes),
+        cores=data_dict.get(MgmtDB.col_cores),
+        memory=data_dict.get(MgmtDB.col_memory),
+        wct=data_dict.get(MgmtDB.col_wct),
```
Comment (jpa198) [inline workflow/automation/execution_scripts/queue_monitor.py:69]: This makes me a little sad, but I also think trying to do val=dict.get(col_val) programmatically would be a bit overkill

File: workflow/automation/lib/schedulers/pbs.py [hunk 3]
```diff
@@ -173,6 +174,30 @@ def check_queues(self, user: bool = False, target_machine=None) -> List[str]:
...
+        limit_time = timedelta(
+            hours=int(limit_hour), minutes=int(limit_min), seconds=int(limit_sec)
+        )
+        _, elapsed_hour, elapsed_min, elapsed_sec = (
+            walltime_split[0].replace('"', "").split(":")
+        )
+        elapsed_time = timedelta(
+            hours=int(elapsed_hour), minutes=int(elapsed_min), seconds=int(elapsed_sec)
+        )
+        return elapsed_time > limit_time
```
Comment (jpa198) [inline workflow/automation/lib/schedulers/pbs.py:199]: If it's equal it also hit WCT. In the case it finished exactly on time, I think most jobs have a short circuit to mark them as done once they start.

File: workflow/automation/execution_scripts/queue_monitor.py [hunk 4]
```diff
@@ -153,12 +160,27 @@ def update_tasks(
                     f"not found on {Scheduler.get_scheduler().QUEUE_NAME} or in the management db folder; resetting the status "
                     "to 'created' for resubmission"
                 )
+
+                # Check for if task was killed by Wall Clock Time
+                try:
+                    killed_wct = Scheduler.get_scheduler().check_wct(
+                        db_running_task.job_id
+                    )
+                except:
```
Comment (jpa198) [inline workflow/automation/execution_scripts/queue_monitor.py:169]: Exception should be narrowed to the types we're expecting, this too broad

File: workflow/automation/lib/schedulers/abstractscheduler.py [hunk 5]
```diff
@@ -70,6 +70,15 @@ def check_queues(self, user: bool = False, target_machine=None) -> List[str]:
         """
         pass
 
+    @abstractmethod
+    def check_wct(self, job_id: int) -> bool:
```
Comment (jpa198) [inline workflow/automation/lib/schedulers/abstractscheduler.py:74]: Name should be 'check_wct_hit' or something and the return should line up with the function name
I also think there should be a third state for undetermined, but None is already a bit loaded as a return value

File: workflow/automation/lib/MgmtDB.py [hunk 6]
```diff
@@ -4,18 +4,29 @@
...
 Process = const.ProcessType
 
-SchedulerTask = namedtuple(
-    "SchedulerTask", ["run_name", "proc_type", "status", "job_id", "error"]
-)
-# Make error an optional value, once we are using Python 3.7 then this can be made nicer..
-SchedulerTask.__new__.__defaults__ = (None,)
+
+@dataclass
+class SchedulerTask:
```
Comment (jpa198) [inline workflow/automation/lib/MgmtDB.py:17]: Do we want to give this class to/from json methods, so the only place json is read/written is in here?

**Review by jpa198 — CHANGES_REQUESTED** [review]
A couple of things, most of them are discussion points rather than must haves though

File: workflow/automation/execution_scripts/queue_monitor.py (hunk 2 above)
Comment (claudio525) [reply workflow/automation/execution_scripts/queue_monitor.py:69]: Use a dict comprehension and then unpack it when passing to init?

File: workflow/automation/org/kisti/vm_pert.pbs [hunk 7]
```diff
@@ -61,5 +62,5 @@ if [[ $pass == 0 ]]; then
 else
     #reformat $res to remove '\n'
     res=`echo $res | tr -d '\n'`
-    python $gmsim/workflow/workflow/automation/execution_scripts/add_to_mgmt_queue.py $MGMT_DB_LOC/mgmt_db_queue $REL_NAME VM_PERT failed $SLURM_JOB_ID --error "$res"
+    python $gmsim/workflow/workflow/automation/execution_scripts/add_to_mgmt_queue.py $MGMT_DB_LOC/mgmt_db_queue $REL_NAME VM_PERT failed $SLURM_JOB_ID --error "$res" --start_time "$start_time" --end_time "$end_time" --nodes $SLURM_NNODES --cores $SLURM_CPUS_PER_TASK --wct 10:00:00
```
Comment (jasonmotha) [inline workflow/automation/org/kisti/vm_pert.pbs:65]: VM Pert is estimated - at least from memory it is.

File: workflow/automation/org/nesi/vm_pert.sl [hunk 8]
```diff
@@ -65,5 +65,5 @@ if [[ $pass == 0 ]]; then
 else
     #reformat $res to remove '\n'
     res=`echo $res | tr -d '\n'`
-    python $gmsim/workflow/workflow/automation/execution_scripts/add_to_mgmt_queue.py $MGMT_DB_LOC/mgmt_db_queue $REL_NAME VM_PERT failed $SLURM_JOB_ID --error "$res"
+    python $gmsim/workflow/workflow/automation/execution_scripts/add_to_mgmt_queue.py $MGMT_DB_LOC/mgmt_db_queue $REL_NAME VM_PERT failed $SLURM_JOB_ID --error "$res" --start_time "$start_time" --end_time "$end_time" --nodes $SLURM_NNODES --cores $SLURM_CPUS_PER_TASK --wct 10:00:00
```
Comment (jasonmotha) [inline workflow/automation/org/nesi/vm_pert.sl:68]: same as above about estimation

**Review by jasonmotha — CHANGES_REQUESTED** [review]

File: workflow/automation/execution_scripts/add_to_mgmt_queue.py [hunk 9]
```diff
@@ -6,6 +6,14 @@
 import qcore.constants as const
 from workflow.automation.lib.shared_automated_workflow import add_to_queue
 
+
+def convert_time(time: str):
```
Comment (jasonmotha) [inline workflow/automation/execution_scripts/add_to_mgmt_queue.py:10]: Maybe calling it `datestr_to_timestamp`?

Or at least having a docstring with a couple of examples.

**Review by jasonmotha — APPROVED** [review]

**Review by jpa198 — APPROVED** [review]

**Review by claudio525 — APPROVED** [review]

**Pull request merged**